Dockerfile

```diff
@@ -39,109 +49,124 @@ COPY Gulpfile.babel.js /opt/warehouse/src/
+RUN set -eux \
+    rm -f /etc/apt/apt.conf.d/docker-clean; \
+    echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
 
+# Install System level Warehouse requirements, this is done before everything
+# else because these are rarely ever going to change.
+RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
+    --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    set -eux; \
+    apt-get update; \
+    apt-get install --no-install-recommends -y \
+        libpq5 \
+        libxml2 \
+        libxslt1.1 \
+        libcurl4 \
+    ;
+
+# ---------------------------------- BUILD ----------------------------------
 
 # Now we're going to build our actual application, but not the actual production
 # image that it gets deployed into.
-FROM python:3.8.2-slim-buster as build
-
-# Define whether we're building a production or a development image. This will
-# generally be used to control whether or not we install our development and
-# test dependencies.
-ARG DEVEL=no
-
-# To enable Ipython in the development environment set to yes (for using ipython
-# as the warehouse shell interpreter,
-# i.e. 'docker-compose run --rm web python -m warehouse shell --type=ipython')
-ARG IPYTHON=no
+FROM base as build
 
 # Install System level Warehouse build requirements, this is done before
 # everything else because these are rarely ever going to change.
-RUN set -x \
-    && apt-get update \
-    && apt-get install --no-install-recommends -y \
-        build-essential libffi-dev libxml2-dev libxslt-dev libpq-dev libcurl4-openssl-dev libssl-dev \
-        $(if [ "$DEVEL" = "yes" ]; then echo 'libjpeg-dev'; fi)
+RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
+    --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    set -eux; \
+    apt-get update; \
+    apt-get install --no-install-recommends -y \
+        build-essential \
+        libcurl4-openssl-dev \
+        libffi-dev \
+        libpq-dev \
+        libssl-dev \
+        libxml2-dev \
+        libxslt-dev \
+    ;
 
 # We create an /opt directory with a virtual environment in it to store our
 # application in.
-RUN set -x \
-    && python3 -m venv /opt/warehouse
-
+RUN python3 -m venv /opt/warehouse
 
-# Now that we've created our virtual environment, we'll go ahead and update
-# our $PATH to refer to it first.
-ENV PATH="/opt/warehouse/bin:${PATH}"
-
-# Next, we want to update pip, setuptools, and wheel inside of this virtual
-# environment to ensure that we have the latest versions of them.
-# TODO: We use --require-hashes in our requirements files, but not here, making
-#       the ones in the requirements files kind of a moot point. We should
-#       probably pin these too, and update them as we do anything else.
-RUN pip --no-cache-dir --disable-pip-version-check install --upgrade pip setuptools wheel
+# Pip configuration (https://github.com/pypa/warehouse/pull/4584)
+ENV PIP_NO_BINARY=hiredis PIP_DISABLE_PIP_VERSION_CHECK=1
```

Comment:
It completely escapes me why we do not use the binary releases for hiredis... but based on the date it seems to be related to https://github.com/redis/hiredis-py/issues/69 (nice) which was seemingly only related to that 0.2.0 release.

Perhaps we should go back to using the wheels provided?

Reply:
Will address in follow up PR

Reply:
:+1: